presto-spi/src/main/java/com/facebook/presto/spi/Page.java

```diff
@@ -146,14 +150,84 @@ public void compact()
         for (int i = 0; i < blocks.length; i++) {
             Block block = blocks[i];
             if (block instanceof DictionaryBlock) {
-                DictionaryId sourceId = ((DictionaryBlock) block).getDictionarySourceId();
-                relatedDictionaryBlocks.computeIfAbsent(sourceId, id -> new DictionaryBlockIndexes())
-                        .addBlock(block, i);
+                DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+                relatedDictionaryBlocks.computeIfAbsent(dictionaryBlock.getDictionarySourceId(), id -> new DictionaryBlockIndexes())
+                        .addBlock(dictionaryBlock, i);
             }
         }
         return relatedDictionaryBlocks;
     }
 
+    private static List<DictionaryBlock> compactRelatedBlocks(List<DictionaryBlock> blocks)
+    {
+        DictionaryBlock firstDictionaryBlock = blocks.get(0);
+        Block dictionary = firstDictionaryBlock.getDictionary();
+        Slice dictionaryIds = firstDictionaryBlock.getIds();
+
+        int positionCount = firstDictionaryBlock.getPositionCount();
+        int dictionarySize = dictionary.getPositionCount();
+
+        // determine which dictionary entries are referenced and build a reindex for them
+        List<Integer> dictionaryPositionsToCopy = new ArrayList<>(dictionarySize);
+        int[] remapIndex = new int[dictionarySize];
+        Arrays.fill(remapIndex, -1);
+
+        boolean[] copied = new boolean[dictionarySize];
+        int newIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = getIndex(dictionaryIds, i);
+            if (!copied[position]) {
+                dictionaryPositionsToCopy.add(position);
+                remapIndex[position] = newIndex;
+                newIndex++;
+                copied[position] = true;
+            }
+        }
+
+        // entire dictionary is referenced
+        if (dictionaryPositionsToCopy.size() == dictionarySize) {
+            return blocks;
+        }
+
+        // compact the dictionaries
+        Slice newIdsSlice = wrappedIntArray(getNewIds(positionCount, dictionaryIds, remapIndex));
+        List<DictionaryBlock> outputDictionaryBlocks = new ArrayList<>(blocks.size());
+        DictionaryId newDictionaryId = randomDictionaryId();
+        for (DictionaryBlock dictionaryBlock : blocks) {
+            if (!firstDictionaryBlock.getDictionarySourceId().equals(dictionaryBlock.getDictionarySourceId())) {
+                throw new IllegalArgumentException("dictionarySourceIds must be the same");
+            }
+
+            try {
+                Block compactDictionary = dictionaryBlock.getDictionary().copyPositions(dictionaryPositionsToCopy);
+                outputDictionaryBlocks.add(new DictionaryBlock(positionCount, compactDictionary, newIdsSlice, true, newDictionaryId));
```

Comment:
I understand this, but each of the `DictionaryBlock`s gets the same references `newIdsSlice`. This confuses me.